include/nbl/builtin/hlsl/bxdf/transmission/ggx.hlsl

```diff
@@ -0,0 +1,422 @@
+        using scalar_type = scalar_type;
+
+        scalar_type getDevshV() NBL_CONST_MEMBER_FUNC { return devsh_v; }
+        scalar_type getDevshL() NBL_CONST_MEMBER_FUNC { return devsh_l; }
+
+        scalar_type devsh_v;
+        scalar_type devsh_l;
+    };
+    using query_type = SGGXQuery;
+
+    static this_t create(NBL_CONST_REF_ARG(fresnel::OrientedEtas<monochrome_type>) orientedEta, scalar_type ax, scalar_type ay)
+    {
+        this_t retval;
+        retval.__base.ndf.A = vector2_type(ax, ay);
+        retval.__base.ndf.ax2 = ax*ax;
+        retval.__base.ndf.ay2 = ay*ay;
+        retval.__base.ndf.a2 = ax*ay;
+        retval.__base.fresnel.orientedEta = orientedEta;
+        retval.__base.fresnel.orientedEta2 = orientedEta.value * orientedEta.value;
+        return retval;
+    }
+    static this_t create(NBL_CONST_REF_ARG(creation_type) params)
+    {
+        return create(params.orientedEta, params.ax, params.ay);
+    }
+
+    query_type createQuery(NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction)
+    {
+        query_type query;
+        ndf_type ggx_ndf = __base.ndf;
+        query.devsh_v = ggx_ndf.devsh_part(interaction.getTdotV2(), interaction.getBdotV2(), interaction.getNdotV2());
+        query.devsh_l = ggx_ndf.devsh_part(_sample.getTdotL2(), _sample.getBdotL2(), _sample.getNdotL2());
+        return query;
+    }
+
+    spectral_type eval(NBL_CONST_REF_ARG(query_type) query, NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction, NBL_CONST_REF_ARG(anisocache_type) cache)
+    {
+        struct SGGXG2XQuery
+        {
+            using scalar_type = scalar_type;
+
+            scalar_type getDevshV() NBL_CONST_MEMBER_FUNC { return devsh_v; }
+            scalar_type getDevshL() NBL_CONST_MEMBER_FUNC { return devsh_l; }
+            BxDFClampMode getClampMode() NBL_CONST_MEMBER_FUNC { return _clamp; }
+
+            scalar_type devsh_v;
+            scalar_type devsh_l;
+            BxDFClampMode _clamp;
+        };
+
+        SGGXG2XQuery g2_query;
+        g2_query.devsh_v = query.getDevshV();
+        g2_query.devsh_l = query.getDevshL();
+        g2_query._clamp = _clamp;
+
+        fresnel::OrientedEtas<monochrome_type> orientedEta = __base.fresnel.orientedEta;
+        measure_transform_type dualMeasure = __base.template __DG<SGGXG2XQuery>(g2_query, _sample, interaction, cache);
+        dualMeasure.absNdotL = _sample.getNdotL(_clamp);
+        dualMeasure.orientedEta = orientedEta.value[0];
+        scalar_type DG = dualMeasure.getProjectedLightMeasure();
+
+        return hlsl::promote<spectral_type>(__base.fresnel(hlsl::abs(cache.getVdotH()))[0]) * DG;
+    }
+
+    sample_type __generate_wo_clamps(const vector3_type localV, const vector3_type H, NBL_CONST_REF_ARG(matrix3x3_type) m, NBL_REF_ARG(vector3_type) u, NBL_CONST_REF_ARG(fresnel::OrientedEtaRcps<monochrome_type>) rcpEta, NBL_REF_ARG(anisocache_type) cache)
```

Comment:
you need to take whole ray-dir instead of vec3 `localV` (same comment for all cook torrance)